### ISSUE TITLE
gorelease cannot be built from source due to missing dependencies

### ISSUE THREAD

**Issue by glaubitz** [goreleaser/goreleaser#1013]
I am currently trying to build ```goreleaser``` which is required for ```aws-iam-authenticator```.

Unfortunately, ```goreleaser``` is currently not buildable due to some of the dependencies not being installable:

```
glaubitz@suse-laptop:~> go get github.com/goreleaser/goreleaser       
package github.com/goreleaser/goreleaser
        imports context: unrecognized import path "context"
package github.com/goreleaser/goreleaser
        imports github.com/Masterminds/semver
        imports github.com/apex/log
        imports github.com/pkg/errors
        imports github.com/apex/log/handlers/cli
        imports github.com/fatih/color
        imports github.com/mattn/go-colorable
        imports github.com/mattn/go-isatty
        imports golang.org/x/sys/unix
        imports github.com/aws/aws-sdk-go/aws
        imports github.com/aws/aws-sdk-go/aws/credentials
        imports github.com/aws/aws-sdk-go/aws/credentials/stscreds
        imports github.com/jmespath/go-jmespath
        imports github.com/aws/aws-sdk-go/aws/session
        imports github.com/aws/aws-sdk-go/service/s3
        imports github.com/caarlos0/ctrlc
        imports github.com/campoy/unique
        imports github.com/google/go-github/v25/github
        imports github.com/google/go-github/v25/github
        imports github.com/google/go-github/v25/github: cannot find package "github.com/google/go-github/v25/github" in any of:
        /usr/lib64/go1.4/src/github.com/google/go-github/v25/github (from $GOROOT)
        /home/glaubitz/go/src/github.com/google/go-github/v25/github (from $GOPATH)
package github.com/goreleaser/goreleaser
        imports github.com/Masterminds/semver
        imports github.com/apex/log
        imports github.com/pkg/errors
        imports github.com/apex/log/handlers/cli
        imports github.com/fatih/color
        imports github.com/mattn/go-colorable
        imports github.com/mattn/go-isatty
        imports golang.org/x/sys/unix
        imports github.com/aws/aws-sdk-go/aws
        imports github.com/aws/aws-sdk-go/aws/credentials
        imports github.com/aws/aws-sdk-go/aws/credentials/stscreds
        imports github.com/jmespath/go-jmespath
        imports github.com/aws/aws-sdk-go/aws/session
        imports github.com/aws/aws-sdk-go/service/s3
        imports github.com/caarlos0/ctrlc
        imports github.com/campoy/unique
        imports github.com/google/go-github/v25/github
        imports gopkg.in/yaml.v2
        imports golang.org/x/oauth2
        imports golang.org/x/net/context/ctxhttp
        imports golang.org/x/sync/errgroup
        imports github.com/mattn/go-zglob
        imports github.com/mitchellh/go-homedir
        imports github.com/goreleaser/nfpm
        imports github.com/imdario/mergo
        imports github.com/goreleaser/nfpm/deb
        imports github.com/blakesmith/ar
        imports github.com/goreleaser/nfpm/rpm
        imports github.com/kamilsk/retry/v4
        imports github.com/kamilsk/retry/v4
        imports github.com/kamilsk/retry/v4: cannot find package "github.com/kamilsk/retry/v4" in any of:
        /usr/lib64/go1.4/src/github.com/kamilsk/retry/v4 (from $GOROOT)
        /home/glaubitz/go/src/github.com/kamilsk/retry/v4 (from $GOPATH)
package github.com/goreleaser/goreleaser
        imports github.com/Masterminds/semver
        imports github.com/apex/log
        imports github.com/pkg/errors
        imports github.com/apex/log/handlers/cli
        imports github.com/fatih/color
        imports github.com/mattn/go-colorable
        imports github.com/mattn/go-isatty
        imports golang.org/x/sys/unix
        imports github.com/aws/aws-sdk-go/aws
        imports github.com/aws/aws-sdk-go/aws/credentials
        imports github.com/aws/aws-sdk-go/aws/credentials/stscreds
        imports github.com/jmespath/go-jmespath
        imports github.com/aws/aws-sdk-go/aws/session
        imports github.com/aws/aws-sdk-go/service/s3
        imports github.com/caarlos0/ctrlc
        imports github.com/campoy/unique
        imports github.com/google/go-github/v25/github
        imports gopkg.in/yaml.v2
        imports golang.org/x/oauth2
        imports golang.org/x/net/context/ctxhttp
        imports golang.org/x/sync/errgroup
        imports github.com/mattn/go-zglob
        imports github.com/mitchellh/go-homedir
        imports github.com/goreleaser/nfpm
        imports github.com/imdario/mergo
        imports github.com/goreleaser/nfpm/deb
        imports github.com/blakesmith/ar
        imports github.com/goreleaser/nfpm/rpm
        imports github.com/kamilsk/retry/v4
        imports github.com/kamilsk/retry/v4/backoff
        imports github.com/kamilsk/retry/v4/backoff
        imports github.com/kamilsk/retry/v4/backoff: cannot find package "github.com/kamilsk/retry/v4/backoff" in any of:
        /usr/lib64/go1.4/src/github.com/kamilsk/retry/v4/backoff (from $GOROOT)
        /home/glaubitz/go/src/github.com/kamilsk/retry/v4/backoff (from $GOPATH)
package github.com/goreleaser/goreleaser
        imports github.com/Masterminds/semver
        imports github.com/apex/log
        imports github.com/pkg/errors
        imports github.com/apex/log/handlers/cli
        imports github.com/fatih/color
        imports github.com/mattn/go-colorable
        imports github.com/mattn/go-isatty
        imports golang.org/x/sys/unix
        imports github.com/aws/aws-sdk-go/aws
        imports github.com/aws/aws-sdk-go/aws/credentials
        imports github.com/aws/aws-sdk-go/aws/credentials/stscreds
        imports github.com/jmespath/go-jmespath
        imports github.com/aws/aws-sdk-go/aws/session
        imports github.com/aws/aws-sdk-go/service/s3
        imports github.com/caarlos0/ctrlc
        imports github.com/campoy/unique
        imports github.com/google/go-github/v25/github
        imports gopkg.in/yaml.v2
        imports golang.org/x/oauth2
        imports golang.org/x/net/context/ctxhttp
        imports golang.org/x/sync/errgroup
        imports github.com/mattn/go-zglob
        imports github.com/mitchellh/go-homedir
        imports github.com/goreleaser/nfpm
        imports github.com/imdario/mergo
        imports github.com/goreleaser/nfpm/deb
        imports github.com/blakesmith/ar
        imports github.com/goreleaser/nfpm/rpm
        imports github.com/kamilsk/retry/v4
        imports github.com/kamilsk/retry/v4/backoff
        imports github.com/kamilsk/retry/v4/strategy
        imports github.com/kamilsk/retry/v4/strategy
        imports github.com/kamilsk/retry/v4/strategy: cannot find package "github.com/kamilsk/retry/v4/strategy" in any of:
        /usr/lib64/go1.4/src/github.com/kamilsk/retry/v4/strategy (from $GOROOT)
        /home/glaubitz/go/src/github.com/kamilsk/retry/v4/strategy (from $GOPATH)
glaubitz@suse-laptop:~>
```

**Comment by caarlos0** [comment]
yeah I don't think this is supported, as it does not use go modules it seems...

please follow one of https://goreleaser.com/install/

thanks!

**Comment by glaubitz** [comment]
I'm just seeing this here: https://goreleaser.com/contributing/

The background is that I am packaging aws-iam-authenticator for openSUSE/SLE, so I have to be able to build the packages from source.

**Comment by caarlos0** [comment]
yeah, not that you don't `go get`, but `git clone`

`make setup` does the `go get` with go modules enabled and etc..

**Comment by glaubitz** [comment]
Not looking at an issue fixes a bug automatically?

**Comment by glaubitz** [comment]
And running ```make setup``` downloads scripts with ```curl``` and pipes them into ```sh``` which is a security nightmare:

```
curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh
golangci/golangci-lint info checking GitHub for latest tag
golangci/golangci-lint info found version: 1.16.0 for v1.16.0/linux/amd64
golangci/golangci-lint info installed ./bin/golangci-lint
curl -sfL https://install.goreleaser.com/github.com/gohugoio/hugo.sh | sh
gohugoio/hugo info checking GitHub for latest tag
gohugoio/hugo info found version: 0.55.5 for v0.55.5/Linux/64bit
gohugoio/hugo info installed ./bin/hugo
go mod download
(...)
```

**Comment by caarlos0** [comment]
well, you can always do the same thing manually...

**Comment by glaubitz** [comment]
I'm going to hack my own build system for ```aws-iam-authenticator``` now. I cannot use ```goreleaser``` for a package that is supposed to be shipped with an enterprise distribution.

**Comment by caarlos0** [comment]
that's OK... you could also just use the distributed goreleaser binaries...

**Comment by caarlos0** [comment]
FWIW you don't even need all those things if you just want to build goreleaser from source:

```
cd /tmp
git clone https://github.com/goreleaser/goreleaser
cd goreleaser
go get ./...
go build -o goreleaser .
./goreleaser --version
```

**Comment by glaubitz** [comment]
I cannot import binaries, everything has to be built from source and without an internet connection.

I don't know of any Linux distribution that allows the import of binaries compiled upstream. This would also cause trouble for other architectures besides x86_64 like ARM, POWER or S390x.

**Comment by caarlos0** [comment]
got it, so steps above should work... you just need to clone outside `GOPATH` or set `GO111MODULE=on` because it uses go modules, and `GO111MODULE` defaults to `auto`, which is `on` inside `GOPATH`...

**Comment by glaubitz** [comment]
Turns out I don't need ```goreleaser```. I'm just iterating over the list of source files in ```aws-iam-authenticator``` and then pipe that into ```go build```, see:

> https://build.opensuse.org/package/show/home:glaubitz:branches:Cloud:Tools/aws-iam-authenticator

**Comment by caarlos0** [comment]
ah, makes sense.

ok then, will close this, feel free to comment here if you think its needed.

**Comment by timbaileyjones** [comment]
I had to [remove goreleaser as a dependency from `stelligent/mu`](https://github.com/stelligent/mu/pull/441)  due to this exact issue.  Mu was failing because the `go get` for goreleaser command was failing.  Like @glaubitz, it turns about that mu didn't need the *source* for goreleaser, just the *binary* from the releases site.  So I had to update the README to just tell the user to come here and get the latest binary.

@caarlos0 - I really think you should re-open this ticket, because not being able to build from source is a pretty serious problem.

**Comment by caarlos0** [comment]
the problem is that goreleaser uses go modules, and `go get` defaults to go modules being disable.

```
GO111MODULE=on go get github.com/goreleaser/goreleaser
```

should work.

**Comment by timbaileyjones** [comment]
Thanks for the response, @caarlos0. I only stumbled on the world of different Go packaging methods this week, so I definitely do not know the topic well yet.

If that all it takes build from source, it would be nice for it to be in the README.md.  

Another engineer from Stelligent took`stelligent/mu` went through this module transition a few months ago, but I wasn't involved in that work.  I learned that `GO111MODULE` defaults to `auto` when the source is in GOPATH/src (and when we work on `mu` locally, that IS the case).
